spec/01-unit/26-observability/05-logs_spec.lua

```diff
@@ -0,0 +1,87 @@
+require "kong.tools.utils"
+
+
+describe("Observability/Logs unit tests", function()
+  describe("maybe_push()", function()
+    local o11y_logs, maybe_push, get_request_logs, get_worker_logs
+    local old_ngx, old_kong
+
+    lazy_setup(function()
+      old_ngx = _G.ngx
+      old_kong = _G.kong
+
+      _G.ngx = {
+        config = { subsystem = "http" },
+        ctx = {},
+        DEBUG = ngx.DEBUG,
+        INFO = ngx.INFO,
+        WARN = ngx.WARN,
+      }
+
+      _G.kong = {
+        configuration = {
+          log_level = "info",
+        },
+      }
+
+      o11y_logs = require "kong.observability.logs"
+      maybe_push = o11y_logs.maybe_push
+      get_request_logs = o11y_logs.get_request_logs
+      get_worker_logs = o11y_logs.get_worker_logs
+    end)
+
+    before_each(function()
+      _G.ngx.ctx = {}
+    end)
+
+    lazy_teardown(function()
+      _G.ngx = old_ngx
+      _G.kong = old_kong
+    end)
+
+    it("has no effect when no log line is provided", function()
+      maybe_push(1, ngx.INFO)
+      local worker_logs = get_worker_logs()
+      assert.same({}, worker_logs)
+      local request_logs = get_request_logs()
+      assert.same({}, request_logs)
+    end)
+
+    it("has no effect when log line is empty", function()
+      maybe_push(1, ngx.INFO, "")
+      local worker_logs = get_worker_logs()
+      assert.same({}, worker_logs)
+      local request_logs = get_request_logs()
+      assert.same({}, request_logs)
+    end)
+
+    it("has no effect when log level is lower than the configured value", function()
+      maybe_push(1, ngx.DEBUG, "Don't mind me, I'm just a debug log")
+      local worker_logs = get_worker_logs()
+      assert.same({}, worker_logs)
+      local request_logs = get_request_logs()
+      assert.same({}, request_logs)
+    end)
+
+    it("generates worker-scoped log entries", function()
+      local log_level = ngx.WARN
+      local body = "Careful! I'm a warning!"
+
+      maybe_push(1, log_level, body, true, 123, ngx.null, nil, function()end, { foo = "bar" })
+      local worker_logs = get_worker_logs()
+      assert.equals(1, #worker_logs)
+
+      local logged_entry = worker_logs[1]
+      assert.same(log_level, logged_entry.log_level)
+      assert.matches(body .. "true123nilnilfunction:%s0x%x+table:%s0x%x+", logged_entry.body)
```

Comment:
`ngx.log` only accept `string`, `number`, `table with __tostring`, `boolean`, and `ngx.null`, we don't need to test other types since `ngx.log` will panic.

Reply:
this is unit testing the "maybe_push" function which is not necessarily bound to ngx.log, it could be called to generate OTel-formatted logs outside of the context of ngx.log, see the (draft, in progress: https://github.com/Kong/kong/pull/13329)

Reply:
So the patched `ngx.log` might send OTel logs that not been produced by native `ngx.log`? Am I right?

I'm ok for this behavior. Would mind to add a comment to the patched version of `ngx.log` to notice this behavior for further maintainers?

Reply:
Another question is do we need to ship telemetry logs like `"function: 0x123456"` or `"table: 0x123456"`.

Reply:
sorry, I wasn't clear, I meant that the observability module could be used to generate OTel log entries without logging such lines, i.e. without calling ngx.log at all. What will be acceptable as input can be better documented in the other PR I linked, where this interface will be exposed via the PDK. At the moment there is no requirement about the format of functions and tables, so I think the output of `tostring` is good enough for now.

Reply:
makes sense